### PR TITLE
Assist to flip equality (==) and negated equality (!=) operands.

### LINKED ISSUE #1023: "Flip ==" assist
Add assist to flip operands of `==`

```rust
    fn dyn_eq(&self, other: &dyn Diagnostic) -> bool {
        match other.downcast_ref::<Self>() {
            None => false,
            Some(it) => it == self,
        }
    }
```

==>

```rust
    fn dyn_eq(&self, other: &dyn Diagnostic) -> bool {
        match other.downcast_ref::<Self>() {
            None => false,
            Some(it) => self == it,
        }
    }
```

See https://github.com/rust-analyzer/rust-analyzer/blob/3d7a330f320ac2ebf73df012556d64be9b120445/crates/ra_assists/src/flip_comma.rs for a similar assist. 

Here's the `BinExpr` API which should be useful: https://github.com/rust-analyzer/rust-analyzer/blob/3d7a330f320ac2ebf73df012556d64be9b120445/crates/ra_syntax/src/ast.rs#L600

### PR DESCRIPTION
**Pull request by marcogroppo**
This PR adds an assist to flip the equality operands.

I hope this is the right way to do this (I'm a newbie...)

Fixes #1023.

### CODE REVIEW

File: crates/ra_assists/src/flip_eq_operands.rs [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+pub(crate) fn flip_eq_operands(mut ctx: AssistCtx<impl HirDatabase>) -> Option<Assist> {
+    let expr = ctx.node_at_offset::<BinExpr>()?;
+    let allowed_ops = [BinOp::EqualityTest, BinOp::NegatedEqualityTest];
+    let expr_op = expr.op()?;
+    if !allowed_ops.iter().any(|o| *o == expr_op) {
+        return None;
+    }
+    let node = expr.syntax();
+    let prev = node.first_child()?;
+    let next = node.last_child()?;
```
Comment (matklad) [inline crates/ra_assists/src/flip_eq_operands.rs]: It’s better to use lhs and rhs methods here

File: crates/ra_assists/src/flip_eq_operands.rs [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+use ra_syntax::{
+    ast::{AstNode, BinExpr, BinOp}
+};
+
+use crate::{AssistCtx, Assist, AssistId};
+
+pub(crate) fn flip_eq_operands(mut ctx: AssistCtx<impl HirDatabase>) -> Option<Assist> {
+    let expr = ctx.node_at_offset::<BinExpr>()?;
+    let allowed_ops = [BinOp::EqualityTest, BinOp::NegatedEqualityTest];
+    let expr_op = expr.op()?;
```
Comment (matklad) [inline crates/ra_assists/src/flip_eq_operands.rs]: Let’s also check hat the cursor is exactly on the operator? That is, that ctx.coffset is contained in op.range

File: crates/ra_assists/src/flip_eq_operands.rs [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    if !allowed_ops.iter().any(|o| *o == expr_op) {
+        return None;
+    }
+    let node = expr.syntax();
+    let prev = node.first_child()?;
+    let next = node.last_child()?;
+    ctx.add_action(AssistId("flip_eq_operands"), "flip equality operands", |edit| {
+        edit.target(node.range());
```
Comment (matklad) [inline crates/ra_assists/src/flip_eq_operands.rs]: A more precise target would be the operator itself

File: crates/ra_assists/src/flip_eq_operands.rs (hunk 2 above)
Comment (marcogroppo) [reply crates/ra_assists/src/flip_eq_operands.rs]: How can I get the exact position of the operator? Something like that would be fine?
```rust
let lhs_range = expr.lhs()?.syntax().range();
let rhs_range = expr.rhs()?.syntax().range();
let op_node_range = TextRange::from_to(lhs_range.end(), rhs_range.start());
if ctx.frange.range.is_subrange(&op_node_range) { ... }
```

Comment (matklad) [reply crates/ra_assists/src/flip_eq_operands.rs]: Good question, looks like there's no method to do this. Let's add it!

I think we should rename `BinExpr::op` and `PrefixExpr::op` to `op_kind`, and reuse `op`  method to return a `SyntaxNode`

Comment (marcogroppo) [reply crates/ra_assists/src/flip_eq_operands.rs]: I can try, but for this use case wouldn't actually be better if the assist applied even when there is some white space? For example if the cursor is here: `1<|>   == 2` the assist should be available or not?

Comment (matklad) [reply crates/ra_assists/src/flip_eq_operands.rs]: The problem with assists is that there are usually many of them, so keeping focus zone precise helps in general

**Comment by marcogroppo** [comment]
Ok, now the assist should target only the actual `==` and `!=` operators and `BinExpr::op()` returns a `SyntaxNode`.

**Comment by matklad** [comment]
Awesome, thanks!

bors r+

**Pull request merged**